chore(style-editor): Implement optimistic updates for real time headless rendering

KevinDavilaDotCMS opened the pull request:
https://github.com/user-attachments/assets/8acf9194-6e25-4fa2-bbab-6f93d638cdcd

kevindaviladev committed:
WIP. Need to commit to change to other branch.

kevindaviladev committed:
WIP. refactor(edit-ema): Enhance style editor components with time machine functionality and code cleanup
- Added time machine capabilities to manage state history for optimistic updates in the style editor.
- Cleaned up code formatting and ensured consistent use of semicolons and newlines across various files.

These changes improve the maintainability and functionality of the style editor, allowing for better state management and user experience.

kevindaviladev committed:
refactor(style-editor): Simplify form handling and improve state management

- Refactored the style editor form component to utilize a dedicated form builder service, enhancing form creation and management.
- Removed unused imports and cleaned up code for better readability.
- Updated methods for extracting and updating style properties to streamline functionality and improve performance.
- Ensured consistent handling of form values during rollback operations.

These changes enhance the maintainability and efficiency of the style editor, providing a more robust user experience.

kevindaviladev committed:
refactor(edit-ema): Remove unused test button and update iframe messenger import

- Removed the test button from the edit-ema editor component to streamline the UI.
- Updated the import path for the UveIframeMessengerService to reflect its new location.
- Cleaned up console log statements in the contentlet tools component for better code clarity.
- Adjusted style properties handling in the GraphQL utility functions for consistency.

These changes enhance the maintainability of the code and improve the overall user experience in the edit-ema editor.

kevindaviladev committed:
refactor(edit-ema): Clean up imports and enhance code organization

- Removed unused imports from the edit-ema editor component and style editor form builder service for improved clarity.
- Updated import paths for consistency across components, ensuring better maintainability.
- Streamlined the GraphQL utility functions by adjusting import statements, enhancing code readability.

These changes contribute to a cleaner codebase and facilitate easier future modifications.

kevindaviladev committed:
Added jsdoc

kevindaviladev committed:
Merge branch 'main' into 34085-task-implement-optimistic-updates-in-style-editor-for-real-time-headless-rendering

rjvelazco commented on core-web/libs/portlets/edit-ema/portlet/src/lib/edit-ema-editor/components/dot-uve-palette/components/dot-uve-style-editor-form/dot-uve-style-editor-form.component.spec.ts:
```diff
@@ -210,7 +228,8 @@ describe('DotUveStyleEditorFormComponent', () => {
 
-    describe('schema changes', () => {
+    // TODO: Remove skip when we have the styleProperties in PageAPI response and remove the untracked in $reloadSchemaEffect.
+    xdescribe('schema changes', () => {
```
Just a comment as a reminder

rjvelazco commented on core-web/libs/portlets/edit-ema/portlet/src/lib/edit-ema-editor/components/dot-uve-palette/components/dot-uve-style-editor-form/dot-uve-style-editor-form.component.ts:
```diff
@@ -41,110 +51,193 @@ import { STYLE_EDITOR_FIELD_TYPES } from '../../../../../shared/consts';
+            )
+            .subscribe((formValues) => {
+                this.#saveStyleProperties(formValues);
+            });
```
Since the operators are executed sequentially, you can do this:

```suggestion
        formValueChanges$
            .pipe(
                distinctUntilChanged((prev, curr) => JSON.stringify(prev) === JSON.stringify(curr)),
                tap((formValues) => this.#updateIframeImmediately(formValues)),
                debounceTime(3000),
            )
            .subscribe((formValues) => this.#saveStyleProperties(formValues));
```

rjvelazco commented on core-web/libs/portlets/edit-ema/portlet/src/lib/services/iframe-messenger/uve-iframe-messenger.service.ts:
```diff
@@ -0,0 +1,108 @@
+@Injectable({
+    providedIn: 'root'
+})
+export class UveIframeMessengerService {
```
This is cool!

rjvelazco commented on core-web/libs/portlets/edit-ema/portlet/src/lib/store/features/timeMachine/withTimeMachine.ts:
````diff
@@ -0,0 +1,336 @@
+ * );
+ * ```
+ */
+export function withTimeMachine<T>(options?: TimeMachineOptions) {
````
YOU ARE THE MAN!

KevinDavilaDotCMS commented on core-web/libs/portlets/edit-ema/portlet/src/lib/edit-ema-editor/components/dot-uve-palette/components/dot-uve-style-editor-form/dot-uve-style-editor-form.component.ts:
```diff
@@ -41,110 +51,193 @@ import { STYLE_EDITOR_FIELD_TYPES } from '../../../../../shared/consts';
+        const activeContentlet = this.#uveStore.activeContentlet();
+        const customGraphqlResponse = this.#uveStore.$customGraphqlResponse();
+
+        if (!activeContentlet || !customGraphqlResponse) {
```
That was a test i did, but we dont need it anymore
Removed!

KevinDavilaDotCMS commented on core-web/libs/portlets/edit-ema/portlet/src/lib/edit-ema-editor/components/dot-uve-palette/components/dot-uve-style-editor-form/dot-uve-style-editor-form.component.ts:
```diff
@@ -41,110 +51,193 @@ import { STYLE_EDITOR_FIELD_TYPES } from '../../../../../shared/consts';
+    readonly STYLE_EDITOR_FIELD_TYPES = STYLE_EDITOR_FIELD_TYPES;
+    readonly $previousIndex = signal(-1);
+
+    readonly #rollbackDetectionEffect = effect(() => {
```
Do you mean previousIndex?

I think no, we need as a local variable to compare with the index on the store

kevindaviladev committed:
refactor(style-editor): Enhance style editor form with message service and debounce functionality

- Integrated MessageService for user notifications on style property saves, improving user feedback.
- Implemented a debounce mechanism for form value changes to optimize performance and reduce unnecessary API calls.
- Updated variable names for consistency and clarity, ensuring better maintainability.
- Cleaned up code and imports across the style editor component and related files.

These changes enhance the user experience and performance of the style editor, providing timely feedback and efficient state management.

kevindaviladev committed:
Merge branch 'main' into 34085-task-implement-optimistic-updates-in-style-editor-for-real-time-headless-rendering

KevinDavilaDotCMS committed:
Moved model order to trigger pipeline

KevinDavilaDotCMS committed:
Fixed typo

zJaaal commented on core-web/libs/portlets/edit-ema/portlet/src/lib/edit-ema-editor/components/dot-uve-palette/components/dot-uve-style-editor-form/utils/style-editor-graphql.utils.ts:
```diff
@@ -0,0 +1,104 @@
+          graphql?: {
+              query: string;
+              variables: Record<string, string>;
+          };
```
I think we have a type for this on the Types Library

KevinDavilaDotCMS replied:
Right! changed!

zJaaal commented on core-web/libs/portlets/edit-ema/portlet/src/lib/edit-ema-editor/components/dot-uve-palette/components/dot-uve-style-editor-form/utils/style-editor-graphql.utils.ts:
```diff
@@ -0,0 +1,104 @@
+    );
+
+    return contentlet?.styleProperties || null;
+}
```
Why dont we extract the whole contentlet and then we manage the styleProperties?
